crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,297 @@
+            result_container.replace(execution_result);
+        });
+
+        JoinHandle {
+            handle: Some(handle),
+            signal_tx: Some(signal_tx),
+            execution_result: mutex,
+            ..Default::default()
+        }
+    }
+
+    /// This function will block until the task has exited, returning a success or an error
+    async fn wait(&mut self) -> anyhow::Result<()> {
+        loop {
+            if let Ok(mut result_container) = self.execution_result.try_lock() {
+                let finished = self.handle.as_ref().unwrap().is_finished();
+                if result_container.is_some() {
+                    return result_container.take().unwrap();
+                } else if finished {
+                    return Err(anyhow::anyhow!("task panicked"));
+                }
+            }
+            tokio::time::sleep(std::time::Duration::from_secs_f32(0.05)).await;
+        }
+    }
+
+    /// This function will signal the running task with the specified signal
+    /// This is a non-blocking operation
+    fn signal(&self, signal: Signal) -> anyhow::Result<()> {
+        if let Some(tx) = &self.signal_tx {
+            let copy = tx.clone();
+            tokio::spawn(async move {
+                let _ = copy.send(signal).await;
+            });
+        }
+        return anyhow::Ok(());
+    }
+
+    async fn abort(&mut self) {
+        self.handle.as_ref().unwrap().abort();
+        let _ = self.wait().await;
+    }
+}
+
+impl Future for JoinHandle {
+    type Output = anyhow::Result<()>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        return match &self.execution_result.try_lock() {
+            Ok(result_container) => {
+                let finished = self.handle.as_ref().unwrap().is_finished();
+                if result_container.is_some() {
+                    let result = result_container.as_ref().unwrap();
+                    if result.is_ok() {
+                        Poll::Ready(anyhow::Ok(()))
+                    } else {
+                        Poll::Ready(Err(anyhow::anyhow!(result
+                            .as_ref()
+                            .unwrap_err()
+                            .to_string())))
+                    }
+                } else if finished {
+                    Poll::Ready(Err(anyhow::anyhow!("task panicked")))
+                } else {
+                    let waker = cx.waker().clone();
+                    tokio::spawn(async move {
+                        tokio::time::sleep(std::time::Duration::from_secs_f32(0.3)).await;
+                        waker.wake();
+                    });
+                    Poll::Pending
+                }
+            }
+            Err(_) => Poll::Pending,
+        };
+    }
+}
+
+// On drop, send shutdown signal and wait for the task to terminate
+impl Drop for JoinHandle {
+    fn drop(&mut self) {
+        if !self.dropped {
+            let mut this = JoinHandle::default();
```

Comment:
shouldn't we set `self.dropped = true` before and after the swap to prevent double drop?

```
if !self.dropped {
   self.dropped = true
   ... 
   std::mem::swap(&mut this, self);

   self.dropped = true;
```